File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.cpp
class Solution {
public:
string longestPalindrome(string s) {
    int n = s.length();
    if (n == 0) return "";

    vector<vector<bool>> dp(n, vector<bool>(n, false));
    int maxLen = 1;
    int start = 0;

    // Every single character is a palindrome
    for (int i = 0; i < n; i++)
        dp[i][i] = true;

    // Check for two-character palindromes
    for (int i = 0; i < n - 1; i++) {
        if (s[i] == s[i + 1]) {
            dp[i][i + 1] = true;
            start = i;
            maxLen = 2;
        }
    }

    // Expand for longer substrings
    for (int len = 3; len <= n; len++) { // Length of substring
        for (int i = 0; i <= n - len; i++) { // Start index
            int j = i + len - 1; // End index
            if (s[i] == s[j] && dp[i + 1][j - 1]) {
                dp[i][j] = true;
                start = i;
                maxLen = len;
            }
        }
    }

    return s.substr(start, maxLen);
}
};/*
Time Complexity: O(n^2): We iterate over all substrings (nested loops).
Space Complexity: O(n^2)Due to the DP table storing results.*/

File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.js
var longestPalindrome = function(s) {
    if (s.length < 2) return s;

    let start = 0;
    let maxLen = 0;

    const expand = (left, right) => {
        while (left >= 0 && right < s.length && s[left] === s[right]) {
            left--;
            right++;
        }
        return right - left - 1; // length of palindrome
    };

    for (let i = 0; i < s.length; i++) {
        const len1 = expand(i, i);       // odd length
        const len2 = expand(i, i + 1);   // even length

        const len = Math.max(len1, len2);

        if (len > maxLen) {
            maxLen = len;
            start = i - Math.floor((len - 1) / 2);
        }
    }

    return s.substring(start, start + maxLen);
};